Added JSQAudioMediaItem class

eliburke opened the pull request:
Added a test case for the JSAAudioMediaItem class
Added a sample audio file to the demo project
Added a sample audio message to the demo project
Added a UIImage category to return the default pause image
Added the JSQAudioMediaItem header to the main header file
Added pause image (Creative Commons License) from NounProject:
https://thenounproject.com/term/pause/183517/

jessesquires commented:
@eliburke awesome! :tada: 

Glanced through this. First impressions: :+1: 

I'll have a bit more feedback soon, within the next couple days. :smile:

jessesquires commented:
Also -- is [this](https://twitter.com/eliburke) you? :smile:

eliburke commented:
Yep, that's me. As you can see from the timeline I don't spend much time on twitter. 😜

I haven't payed a lot of attention to how you make the sausage, but if you need me to fix bugs like the Travis CI warning, or tweak the API / property names, or submit to release_8.0, just let me know.

jessesquires commented:
Hey @eliburke -- sorry for the delay here. I haven't forgotten. Will review soon.

eliburke commented:
No worries.
I've updated my PR to fix the warnings and remove an unused property. I added a new property to allow users to control the playback audio category, and a delegate method to get notification if the audio category changes. My reasoning for this design is:
1) developers who don't care what sound category is set can ignore it. JSQAudioMediaItem will set the category to Playback and the options to Duck, Speaker, Bluetooth. Sound will play through the mute switch and to the selected outputs. 
2) developers who need different sound options can set them
3) anyone who needs to restore the sound category to a particular setting can do so when they clean up the view controller. I didn't want to try and restore the category within JSQAudioMediaItem because it is possible to have multiple audio streams playing at once, and without some sort of singleton to manage the setting, things would likely go haywire. 

If you think it's necessary I could add additional methods for starting and ending playback.

jessesquires commented on JSQMessagesViewController/Model/JSQAudioMediaItem.h:
```diff
@@ -0,0 +1,124 @@
+
+@class JSQAudioMediaItem;
+@protocol JSQAudioMediaItemDelegate <NSObject>
+@optional
```
let's make these non-optional

jessesquires commented on JSQMessagesViewController/Model/JSQAudioMediaItem.h:
```diff
@@ -0,0 +1,124 @@
+- (void)audioMediaItem:(JSQAudioMediaItem*)audioMediaItem didChangeOriginalAudioCategory:(NSString *)category originalOptions:(AVAudioSessionCategoryOptions)options;
+
+/* didNotChangeCategory is called if JSQAudioMediaItem fails to change the category or categoryOptions */
+- (void)audioMediaItem:(JSQAudioMediaItem*)audioMediaItem didNotChangeCategory:(NSError*)error;
```
could this be a single method?

```objective-c
audioMediaItem: didChangeAudioCategory: initialOptions: error:
```

jessesquires commented on JSQMessagesViewController/Model/JSQAudioMediaItem.h:
```diff
@@ -0,0 +1,124 @@
+@interface JSQAudioMediaItem : JSQMediaItem <JSQMessageMediaData, AVAudioPlayerDelegate, NSCoding, NSCopying>
+
+/**
+ *  delegate used for notification of audio events
```
> The delegate object for audio event notifications

jessesquires commented on JSQMessagesViewController/Model/JSQAudioMediaItem.h:
```diff
@@ -0,0 +1,124 @@
+/**
+ * CGFloat used to determine padding between the button, progress bar, and label
+ */
+@property (nonatomic) CGFloat controlPadding;
```
Let's bundle up everything from `playButtonImage` to here in a new object, `JSQAudioViewConfiguration`

This class can then own a config and apply it to the view.

jessesquires commented on JSQMessagesViewController/Model/JSQAudioMediaItem.m:
```diff
@@ -0,0 +1,389 @@
+    
+    _audioCategory = @"AVAudioSessionCategoryPlayback";
+    _audioCategoryOptions = AVAudioSessionCategoryOptionDuckOthers | AVAudioSessionCategoryOptionDefaultToSpeaker | AVAudioSessionCategoryOptionAllowBluetooth;
+}
```
We'll need to refine this initializer flow:

- we need a designated init, which should receive all params. this should be the `NSData` init
- the `NSURL` init can delegate to the `NSData` init (with `dataWithContentsOfURL:...`)
- implement `-init` with all defaults instead of `defaultsInit` (but `-init` should call the designated init)

jessesquires commented on JSQMessagesViewController/Model/JSQAudioMediaItem.m:
```diff
@@ -0,0 +1,389 @@
+    _cachedMediaView = nil;
+}
+
+#pragma mark - Internal logic
```
nit: `#pragma mark - Private`

jessesquires commented on JSQMessagesViewController/Model/JSQAudioMediaItem.m:
```diff
@@ -0,0 +1,389 @@
+#pragma mark - Internal logic
+
+- (NSString *)timestampString:(NSTimeInterval)currentTime forDuration:(NSTimeInterval)duration {
+    // print the time as ss.x up to 59 seconds
```
let's do `0:ss` for < 1 min :smile:

jessesquires commented on JSQMessagesViewController/Model/JSQAudioMediaItem.m:
```diff
@@ -0,0 +1,389 @@
+            if (error && [self.delegate respondsToSelector:@selector(audioMediaItem:didNotChangeCategory:)]) {
+                [self.delegate audioMediaItem:self didNotChangeCategory:error];
+            } else if ([self.delegate respondsToSelector:@selector(audioMediaItem:didChangeOriginalAudioCategory:originalOptions:)]) {
+                [self.delegate audioMediaItem:self didChangeOriginalAudioCategory:category originalOptions:options];
```
since we're removing `@optional`, we can remove `respondsToSelector:` checks

jessesquires commented on JSQMessagesViewController/Model/JSQAudioMediaItem.m:
```diff
@@ -0,0 +1,389 @@
+        
+        // fade the button from play to pause
+        [UIView transitionWithView:_playButton
+                          duration:.3
```
`.3` seems a bit long. maybe `.2` ?

jessesquires commented on JSQMessagesViewController/Model/JSQAudioMediaItem.m:
```diff
@@ -0,0 +1,389 @@
+                       successfully:(BOOL)flag {
+    
+    [_progressTimer invalidate];
+    _progressTimer = nil;
```
these 2 lines are in multiple places, let's pull out into a new helper method

jessesquires commented on JSQMessagesViewController/Model/JSQAudioMediaItem.m:
```diff
@@ -0,0 +1,389 @@
+    }
+    if (! self.pauseButtonImage) {
+        self.pauseButtonImage = [[UIImage jsq_defaultPauseImage] jsq_imageMaskedWithColor:self.tintColor];
+    }
```
the `DisplaySize` method seems like an odd place to initialize these for the first time...

could this happen in `-mediaView` instead?

jessesquires commented on JSQMessagesViewController/Model/JSQAudioMediaItem.m:
```diff
@@ -0,0 +1,389 @@
+        [playView addSubview:_progressView];
+
+        [JSQMessagesMediaViewBubbleImageMasker applyBubbleImageMaskToMediaView:playView isOutgoing:self.appliesMediaViewMaskAsOutgoing];
+        self.cachedMediaView = playView;
```
Let's pull all of this view construction out into a new class, `JSQAudioMediaView`

this should have an `applyConfiguration:` method to do with the config class described above.

jessesquires commented on JSQMessagesViewController/Model/JSQAudioMediaItem.m:
```diff
@@ -0,0 +1,389 @@
+    if ((self.audioURL && [self.audioURL isEqual:audioItem.audioURL]) ||
+        (self.audioData && [self.audioData isEqualToData:audioItem.audioData]))
+        return self.isReadyToPlay == audioItem.isReadyToPlay;
+    else return NO;
```
- remove the `else`
- wrap `if` in `{ }`

jessesquires commented on JSQMessagesViewController/Model/JSQAudioMediaItem.m:
```diff
@@ -0,0 +1,389 @@
+{
+    if (self.audioURL)
+        return super.hash ^ self.audioURL.hash;
+    else return super.hash ^ self.audioData.hash;
```
same here:

```objective-c
if (self.audioURL) {
    return super.hash ^ self.audioURL.hash;
}

return super.hash ^ self.audioData.hash;
```

jessesquires commented:
Thanks @eliburke ! Finally left some comments. Mostly organization stuff.

One other concern:

Can we simplify by *only* storing the `NSData` **or** the `NSURL` and provide a computed property for the one we don't keep? 

Actually, I don't think we can go `NSData` -> `NSURL`, but only `NSURL` -> `NSData`. 

This will help simplify things a lot if we only have to worry about `NSData` internally.

JOMEN5 commented:
looking forward to your guys update for JSQAudioMediaItem,I can't wait to use it.

eliburke commented:
@JOMEN5 I'm on vacation with the family right now. Hard to find time for coding between kids and rum punch. I'll try to get an updated PR in the next couple of days.

JOMEN5 commented:
@eliburke It's ok,family is important,enjoy your vacation with your lovely kids and say hello to them for me.By the way,@jessesquires also should find some time to take a rest,don't put much pressure on yourself.This library is fantastic,thank you again for your's contribution and with my best wishes!

eliburke commented:
@jessesquires I've made the majority of your changes: 
- Collapsed the protocol to one non-optional method
- Added a view config class
- added a designated init and various convenience init methods
- updated the timestamp format to match iMessages
- updated various comments and pragmas
- updated the fade in/out timer
- added timer manipulation methods
- removed the internal audio URL and isReadyToPlay properties, replaced with a convenience setter that calls NSData:dataWithContentsOfURL

The one change I did not make was moving the mediaView logic out into a separate class. To be frank, I don't see any value, just extra complexity. If you are thinking that this re-design would allow someone to provide their own view class that plugs into the audio media item, all of the avAudioPlayer controls and the timer would have would have to be shimmed into the new class with yet another protocol. Or, if you put the avAudioPlayer and timer into the new view class, the audio media item class becomes an empty container. 

If someone REALLY wants to revamp the look of the audio player beyond the viewConfig settings, I think it would be far simpler to copy and modify the base implementation to suit.

JOMEN5 commented:
@eliburke Thanks for your contribution! I'm here also remain @jessesquires update this PR when you have free time. Best wishes.

jessesquires commented on JSQMessagesViewController/Model/JSQAudioMediaItem.h:
```diff
@@ -0,0 +1,153 @@
+
+@end
+
+@interface JSQAudioMediaViewConfiguration : NSObject
```
let's put this new class in its own file :smile:

jessesquires commented on JSQMessagesViewController/Model/JSQAudioMediaItem.m:
```diff
@@ -0,0 +1,368 @@
+@property (strong, nonatomic) UILabel *         progressLabel;
+@property (strong, nonatomic) NSTimer *         progressTimer;
+
+@property (strong, nonatomic) AVAudioPlayer *   audioPlayer;
```
let's follow existing style to be consistent:

```
// var name next to *
@property (strong, nonatomic) UIView *cachedMediaView;
```

jessesquires commented on JSQMessagesViewController/Model/JSQAudioMediaItem.m:
```diff
@@ -0,0 +1,368 @@
+    if (duration < 60) {
+        if (_audioViewConfiguration.showFractionalSeconds)
+            return [NSString stringWithFormat:@"%.01f", currentTime];
+        else if (currentTime < duration)
```
braces `{ }` around every if / else

jessesquires commented:
Thanks so much @eliburke !

Left a couple more really minor comments. Then I think this is good to go. :+1:

JOMEN5 commented:
Thanks @jessesquires and have you finish the task now? How can I get the latest version,just use pod update or pod 'JSQMessagesViewController', :git => 'https://github.com/jessesquires/JSQMessagesViewController.git', :branch => 'develop'?Looking forward to your reply.

jessesquires commented:
@eliburke let me know when those last couple fixes are pushed and i'll merge.

(or, if travis behaves, merge yourself :smile: )

jessesquires commented:
:tada:
